File: Backend/src/main/java/lan/dk/podcastserver/service/HtmlService.java
package lan.dk.podcastserver.service;

import com.mashape.unirest.http.HttpResponse;
import io.vavr.control.Option;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.springframework.stereotype.Service;

import java.nio.charset.StandardCharsets;
import java.util.stream.Collector;
import static io.vavr.API.*;
/**
 * Created by kevin on 07/06/15 for HackerRank problem
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class HtmlService {

    private final UrlService urlService;

    public Option<Document> get(String url) {
        return Try(() -> urlService.get(url)
                    .header(UrlService.USER_AGENT_KEY, UrlService.USER_AGENT_DESKTOP)
                .asBinary())
                    .filter(h -> h.getStatus() < 400)
                    .map(HttpResponse::getBody)
                    .mapTry(is -> Jsoup.parse(is, StandardCharsets.UTF_8.name(), ""))
                    .onFailure(e -> log.error("Error during HTML Fetching of {}", url, e))
                .toOption();
    }

    public Document parse(String html) {
        return Jsoup.parse(html);
    }

    public static Collector<Element, Elements, Elements> toElements() {
        return Collector.of(
                Elements::new,
                Elements::add,
                (left, right) -> { left.addAll(right); return left; },
                Collector.Characteristics.UNORDERED
        );
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/service/HtmlService.kt
package com.github.davinkevin.podcastserver.service

import arrow.core.Failure
import arrow.core.Try
import com.github.davinkevin.podcastserver.utils.toVΛVΓ
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.jsoup.select.Elements
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.nio.charset.StandardCharsets
import java.util.function.BiConsumer
import java.util.function.BinaryOperator
import java.util.function.Supplier
import java.util.stream.Collector

@Service
class HtmlService(var urlService: UrlService) {

    val log = LoggerFactory.getLogger(this.javaClass.name)!!

    fun get(url: String): io.vavr.control.Option<Document> {

        val r = Try { urlService.get(url).header(UrlService.USER_AGENT_KEY, UrlService.USER_AGENT_DESKTOP).asBinary() }
                .filter { it.status < 400 }
                .map { it.body }
                .flatMap { Try { Jsoup.parse(it, StandardCharsets.UTF_8.name(), "") } }

        if (r is Failure) {
            log.error("Error during HTML Fetching of {}", url, r.exception)
        }

        return r.toOption().toVΛVΓ()
    }

    fun parse(html: String): Document = Jsoup.parse(html)

    companion object {

        @JvmStatic
        fun toElements(): Collector<Element, Elements, Elements> {
            return Collector.of<Element, Elements>(
                    Supplier { Elements() },
                    BiConsumer { obj, e -> obj.add(e) },
                    BinaryOperator { left, right -> left.addAll(right); left },
                    Collector.Characteristics.UNORDERED
            )
        }
    }
}